### ISSUE TITLE
Implement schema caching

### ISSUE THREAD

**Issue by TylerBarnes** [wp-graphql/wp-graphql#1160]
Currently a schema that is too large may result in timeouts. This seems to be very rare but can happen.

One proposed solution Jason had is to do what Apollo server does: generate and cache parts of the schema individually and then combine them together. On every request the individual caches could be regenerated async.

**Comment by thefrosty** [comment]
Opened after I saw timeout errors when adding 19 custom post types and then loading the complete schema via web graphiql. Saw the following error plus additional errors throughout the `GraphQL\Executor\Executor` class.
```
PHP Fatal error:  Maximum execution time of 30 seconds exceeded 
in /../wp-content/plugins/wp-graphql/vendor/webonyx/graphql-php/src/Utils/MixedStore.php on line 150
```
Changing my local max_execution_time to 90 worked, but not sure that's a _proper_ solution.

**Comment by TylerBarnes** [comment]
I just ran into this as well with a fairly minimal ACF setup that had a couple repeaters, a flex field, and 2 relationship fields. 

```json
[
    {
        "key": "group_5e5d71e2746ef",
        "title": "Page",
        "fields": [
            {
                "key": "field_5e5d71e8712f4",
                "label": "Image",
                "name": "image",
                "type": "image",
                "instructions": "",
                "required": 0,
                "conditional_logic": 0,
                "wrapper": {
                    "width": "",
                    "class": "",
                    "id": ""
                },
                "show_in_graphql": 1,
                "return_format": "array",
                "preview_size": "medium",
                "library": "all",
                "min_width": "",
                "min_height": "",
                "min_size": "",
                "max_width": "",
                "max_height": "",
                "max_size": "",
                "mime_types": ""
            },
            {
                "key": "field_5e5d71ef712f5",
                "label": "flex",
                "name": "flex",
                "type": "flexible_content",
                "instructions": "",
                "required": 0,
                "conditional_logic": 0,
                "wrapper": {
                    "width": "",
                    "class": "",
                    "id": ""
                },
                "show_in_graphql": 1,
                "layouts": {
                    "layout_5e5d71f279f76": {
                        "key": "layout_5e5d71f279f76",
                        "name": "header",
                        "label": "header",
                        "display": "block",
                        "sub_fields": [
                            {
                                "key": "field_5e5d71f8712f6",
                                "label": "Header",
                                "name": "header",
                                "type": "text",
                                "instructions": "",
                                "required": 0,
                                "conditional_logic": 0,
                                "wrapper": {
                                    "width": "",
                                    "class": "",
                                    "id": ""
                                },
                                "show_in_graphql": 1,
                                "default_value": "",
                                "placeholder": "",
                                "prepend": "",
                                "append": "",
                                "maxlength": ""
                            }
                        ],
                        "min": "",
                        "max": ""
                    },
                    "layout_5e5d71ff712f7": {
                        "key": "layout_5e5d71ff712f7",
                        "name": "projects",
                        "label": "Projects",
                        "display": "block",
                        "sub_fields": [],
                        "min": "",
                        "max": ""
                    },
                    "layout_5e5d88a05d147": {
                        "key": "layout_5e5d88a05d147",
                        "name": "team_members",
                        "label": "Team members",
                        "display": "block",
                        "sub_fields": [
                            {
                                "key": "field_5e5d88a75d148",
                                "label": "team members",
                                "name": "team_members",
                                "type": "repeater",
                                "instructions": "",
                                "required": 0,
                                "conditional_logic": 0,
                                "wrapper": {
                                    "width": "",
                                    "class": "",
                                    "id": ""
                                },
                                "show_in_graphql": 1,
                                "collapsed": "",
                                "min": 0,
                                "max": 0,
                                "layout": "table",
                                "button_label": "",
                                "sub_fields": [
                                    {
                                        "key": "field_5e5d88b05d149",
                                        "label": "team member",
                                        "name": "team_member",
                                        "type": "relationship",
                                        "instructions": "",
                                        "required": 0,
                                        "conditional_logic": 0,
                                        "wrapper": {
                                            "width": "",
                                            "class": "",
                                            "id": ""
                                        },
                                        "show_in_graphql": 1,
                                        "post_type": [
                                            "team_member"
                                        ],
                                        "taxonomy": "",
                                        "filters": [
                                            "search",
                                            "post_type",
                                            "taxonomy"
                                        ],
                                        "elements": "",
                                        "min": "",
                                        "max": "",
                                        "return_format": "object"
                                    }
                                ]
                            }
                        ],
                        "min": "",
                        "max": ""
                    },
                    "layout_5e5d88c95d14a": {
                        "key": "layout_5e5d88c95d14a",
                        "name": "project",
                        "label": "project",
                        "display": "block",
                        "sub_fields": [
                            {
                                "key": "field_5e5d88d15d14b",
                                "label": "project relationship",
                                "name": "project_relationship",
                                "type": "relationship",
                                "instructions": "",
                                "required": 0,
                                "conditional_logic": 0,
                                "wrapper": {
                                    "width": "",
                                    "class": "",
                                    "id": ""
                                },
                                "show_in_graphql": 1,
                                "post_type": [
                                    "project"
                                ],
                                "taxonomy": "",
                                "filters": [
                                    "search",
                                    "post_type",
                                    "taxonomy"
                                ],
                                "elements": "",
                                "min": "",
                                "max": "",
                                "return_format": "object"
                            }
                        ],
                        "min": "",
                        "max": ""
                    }
                },
                "button_label": "Add Row",
                "min": "",
                "max": ""
            }
        ],
        "location": [
            [
                {
                    "param": "post_type",
                    "operator": "==",
                    "value": "page"
                }
            ]
        ],
        "menu_order": 0,
        "position": "normal",
        "style": "default",
        "label_placement": "top",
        "instruction_placement": "label",
        "hide_on_screen": "",
        "active": true,
        "description": "",
        "show_in_graphql": 1,
        "graphql_field_name": "acfPageFields"
    },
    {
        "key": "group_5e5d8616df60a",
        "title": "projects",
        "fields": [
            {
                "key": "field_5e5d8628dde4a",
                "label": "project url",
                "name": "project_url",
                "type": "link",
                "instructions": "",
                "required": 0,
                "conditional_logic": 0,
                "wrapper": {
                    "width": "",
                    "class": "",
                    "id": ""
                },
                "show_in_graphql": 1,
                "return_format": "array"
            },
            {
                "key": "field_5e5d8631dde4b",
                "label": "image",
                "name": "image",
                "type": "image",
                "instructions": "",
                "required": 0,
                "conditional_logic": 0,
                "wrapper": {
                    "width": "",
                    "class": "",
                    "id": ""
                },
                "show_in_graphql": 1,
                "return_format": "array",
                "preview_size": "medium",
                "library": "all",
                "min_width": "",
                "min_height": "",
                "min_size": "",
                "max_width": "",
                "max_height": "",
                "max_size": "",
                "mime_types": ""
            }
        ],
        "location": [
            [
                {
                    "param": "post_type",
                    "operator": "==",
                    "value": "project"
                }
            ]
        ],
        "menu_order": 0,
        "position": "normal",
        "style": "default",
        "label_placement": "top",
        "instruction_placement": "label",
        "hide_on_screen": "",
        "active": true,
        "description": "",
        "show_in_graphql": 1,
        "graphql_field_name": "acfProject"
    },
    {
        "key": "group_5e5d724c5c8ba",
        "title": "team member",
        "fields": [
            {
                "key": "field_5e5d725460853",
                "label": "Name",
                "name": "name",
                "type": "text",
                "instructions": "",
                "required": 0,
                "conditional_logic": 0,
                "wrapper": {
                    "width": "",
                    "class": "",
                    "id": ""
                },
                "show_in_graphql": 1,
                "default_value": "",
                "placeholder": "",
                "prepend": "",
                "append": "",
                "maxlength": ""
            },
            {
                "key": "field_5e5d726060854",
                "label": "Title",
                "name": "title",
                "type": "text",
                "instructions": "",
                "required": 0,
                "conditional_logic": 0,
                "wrapper": {
                    "width": "",
                    "class": "",
                    "id": ""
                },
                "show_in_graphql": 1,
                "default_value": "",
                "placeholder": "",
                "prepend": "",
                "append": "",
                "maxlength": ""
            },
            {
                "key": "field_5e5d727560855",
                "label": "Portrait",
                "name": "portrait",
                "type": "image",
                "instructions": "",
                "required": 0,
                "conditional_logic": 0,
                "wrapper": {
                    "width": "",
                    "class": "",
                    "id": ""
                },
                "show_in_graphql": 1,
                "return_format": "array",
                "preview_size": "medium",
                "library": "all",
                "min_width": "",
                "min_height": "",
                "min_size": "",
                "max_width": "",
                "max_height": "",
                "max_size": "",
                "mime_types": ""
            },
            {
                "key": "field_5e5d73f3f4d05",
                "label": "twitter link",
                "name": "twitterlink",
                "type": "link",
                "instructions": "",
                "required": 0,
                "conditional_logic": 0,
                "wrapper": {
                    "width": "",
                    "class": "",
                    "id": ""
                },
                "show_in_graphql": 1,
                "return_format": "array"
            },
            {
                "key": "field_5e5d740688701",
                "label": "web site",
                "name": "web_site",
                "type": "link",
                "instructions": "",
                "required": 0,
                "conditional_logic": 0,
                "wrapper": {
                    "width": "",
                    "class": "",
                    "id": ""
                },
                "show_in_graphql": 1,
                "return_format": "array"
            },
            {
                "key": "field_5e5d87b185982",
                "label": "projects",
                "name": "projects",
                "type": "relationship",
                "instructions": "",
                "required": 0,
                "conditional_logic": 0,
                "wrapper": {
                    "width": "",
                    "class": "",
                    "id": ""
                },
                "show_in_graphql": 1,
                "post_type": [
                    "project"
                ],
                "taxonomy": "",
                "filters": [
                    "search",
                    "post_type",
                    "taxonomy"
                ],
                "elements": "",
                "min": "",
                "max": "",
                "return_format": "object"
            }
        ],
        "location": [
            [
                {
                    "param": "post_type",
                    "operator": "==",
                    "value": "team_member"
                }
            ]
        ],
        "menu_order": 0,
        "position": "normal",
        "style": "default",
        "label_placement": "top",
        "instruction_placement": "label",
        "hide_on_screen": "",
        "active": true,
        "description": "",
        "show_in_graphql": 0,
        "graphql_field_name": "acfData"
    }
]
```

**Comment by TylerBarnes** [comment]
@jasonbahl we might need to do something about this sooner than later

**Comment by jasonbahl** [comment]
@TylerBarnes weird. . . what is your max_execution_time set to? That doesn't seem like it should take long at all to generate the Schema for just a few ACF fields 🤔

**Comment by jasonbahl** [comment]
@TylerBarnes how many post types and taxonomies do you have as well? I see in your location rules some custom types like `Team Members` and `project`

**Comment by TylerBarnes** [comment]
@jasonbahl I'll put it on my list to add a repro next week. I think my only CPT was team members and this was with Local by flywheel (latest version)

**Comment by jasonbahl** [comment]
@TylerBarnes very strange. I've had some fairly large Schemas in my local by flywheel (I'm on the new lightning version now too) instances with no issues. Wonder if I have some config set differently to give my local more resources. 🤔

**Comment by jasonbahl** [comment]
I think the root cause of this issue was addressed with v1.6 release which significantly improved performance for introspection (see notes here: https://github.com/wp-graphql/wp-graphql/releases/tag/v1.6.0)

Additionally, for caching, there have been some advancements such as https://github.com/wp-graphql/wp-graphql-smart-cache which is solving problems in this space
